### PR TITLE
8301703: java.base jdk.internal.foreign.abi.BindingSpecializer uses ASM to generate classes

### PR DESCRIPTION
**Pull request by JornVernee**
Rewrite BindingSpecializer to use the new class file API.

Note: There is a big try/catch/finally block generated in the `specialize` method that currently uses labels. I looked at replacing this with a call to `CodeBuilder::trying` but it would require threading the nested code builders through all the `emit*` methods, which currently access the 'global' CodeBuilder instance attached to the BindingSpecializer instance. Since there didn't really seem to be a big benefit to this, I've kept that try/catch/finally block as is, using labels.

The current implementation could also use `CheckClassAdapter` to do additional verification on the generated bytecode (ahead of the VM's verifier). I'm not sure if the new API has a replacement for that?

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301703](https://bugs.openjdk.org/browse/JDK-8301703): java.base jdk.internal.foreign.abi.BindingSpecializer uses ASM to generate classes

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [95c8d873](https://git.openjdk.org/jdk/pull/13247/files/95c8d873408cd823b622d9a839ebf78f28de4ba4)
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - Committer) ⚠️ Review applies to [95c8d873](https://git.openjdk.org/jdk/pull/13247/files/95c8d873408cd823b622d9a839ebf78f28de4ba4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13247/head:pull/13247` \
`$ git checkout pull/13247`

Update a local copy of the PR: \
`$ git checkout pull/13247` \
`$ git pull https://git.openjdk.org/jdk.git pull/13247/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13247`

View PR using the GUI difftool: \
`$ git pr show -t 13247`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13247.diff">https://git.openjdk.org/jdk/pull/13247.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13247#issuecomment-1490706234)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java [hunk 1]
```diff
@@ -211,11 +205,6 @@ private static byte[] specializeHelper(MethodType leafType, MethodType callerMet
             }
         }
 
-        if (PERFORM_VERIFICATION) {
-            boolean printResults = false; // only print in case of exception
-            CheckClassAdapter.verify(new ClassReader(bytes), null, printResults, new PrintWriter(System.err));
```
Comment (liach) [inline src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Classfile API provides verification functionality as well, as seen here:
https://github.com/openjdk/jdk/blob/83cf28f99639d80e62c4031c4c9752460de5f36c/test/jdk/jdk/classfile/ClassHierarchyInfoTest.java#L105

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Does this provide additional verification over what is already done just by generating the class?

For instance, IIRC the ASM verifier could catch e.g. stack underflow, but that seems to be caught already by the new implementation without running the verifier.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Yes, for instance, the class generation doesn't check operand stack underflow, that you can generate code with invalid pops. The classfile verifier catches this: https://github.com/openjdk/jdk/blob/83cf28f99639d80e62c4031c4c9752460de5f36c/src/java.base/share/classes/jdk/internal/classfile/impl/verifier/VerificationFrame.java#L153

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: This seems to be caught without the verifier as well:

```
Caused by: java.lang.IllegalStateException: Operand stack underflow at bytecode offset 79 of method invoke(SegmentAllocator,MemorySegment,MemorySegment)
  ...
        at java.base/jdk.internal.classfile.impl.StackMapGenerator.generatorError(StackMapGenerator.java:876)
        at java.base/jdk.internal.classfile.impl.StackMapGenerator.generatorError(StackMapGenerator.java:832)
        at java.base/jdk.internal.classfile.impl.StackMapGenerator$Frame.decStack(StackMapGenerator.java:1024)
        at java.base/jdk.internal.classfile.impl.StackMapGenerator.processBlock(StackMapGenerator.java:600)
        at java.base/jdk.internal.classfile.impl.StackMapGenerator.processMethod(StackMapGenerator.java:420)
        at java.base/jdk.internal.classfile.impl.StackMapGenerator.generate(StackMapGenerator.java:293)
        at java.base/jdk.internal.classfile.impl.StackMapGenerator.<init>(StackMapGenerator.java:232)
        at java.base/jdk.internal.classfile.impl.DirectCodeBuilder$4.writeBody(DirectCodeBuilder.java:333)
        at java.base/jdk.internal.classfile.impl.UnboundAttribute$AdHocAttribute.writeTo(UnboundAttribute.java:914)
        at java.base/jdk.internal.classfile.impl.AttributeHolder.writeTo(AttributeHolder.java:56)
        at java.base/jdk.internal.classfile.impl.DirectMethodBuilder.writeTo(DirectMethodBuilder.java:136)
        at java.base/jdk.internal.classfile.impl.BufWriterImpl.writeList(BufWriterImpl.java:194)
        at java.base/jdk.internal.classfile.impl.DirectClassBuilder.build(DirectClassBuilder.java:176)
        at java.base/jdk.internal.classfile.Classfile.build(Classfile.java:218)
        at java.base/jdk.internal.classfile.Classfile.build(Classfile.java:200)
        at java.base/jdk.internal.classfile.Classfile.build(Classfile.java:186)
        at java.base/jdk.internal.foreign.abi.BindingSpecializer.specializeHelper(BindingSpecializer.java:186)
        ...
```

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Just to clarify: I'm looking for the kind of errors that don't get caught by just generating the class, but are also more informative than the default VerifyError you would get from loading an invalid class.

Comment (liach) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Classfile API does have some sort of error reporting mechanism in verification; at least it tracks the error context.
https://github.com/openjdk/jdk/blob/83cf28f99639d80e62c4031c4c9752460de5f36c/src/java.base/share/classes/jdk/internal/classfile/impl/verifier/VerifierImpl.java#L1810-L1828
The method dumping is providing a yaml tree representation of the class file to the provided Consumer, and the VerifyError itself already includes the error context. What exact specific information are you looking for that's provided by ASM? We might be able to add it to Classfile API too.

Comment (JornVernee) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: I'm not really looking for anything specific. I'm just trying to figure out if it's worth it to keep the `PERFORM_VERIFICATION` flag, and change it to call the verifier in the new impl. i.e. does it catch more errors than just generating and loading the class would (or does it output the errors in a better format).

I had a brief look at the implementation, and it seems that the consumer is for detailed logging of the verification process. I think in this case we're just interested in catching errors.

I already tried switching the code to call `Classfile.parse(newBytes).verify(null).forEach(System.err::println)`, but the error I artificially introduced was already being caught during class generation. So I'm wondering if having a separate verification pass will actually catch any additional errors.

Comment (mcimadamore) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: I believe that, in order to generate the actual bytecodes, the classfile API does a full verification pass (as it needs to infer the stackmap information). This leads me to believe that, yes, most (but probably all) errors would be detected simply by generating code. Maybe @asotona can clarify.

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java [hunk 2]
```diff
@@ -913,166 +891,54 @@ private static String valueLayoutConstantFor(Class<?> type) {
...
+            return CD_ValueLayout_OfDouble;
         } else if (type == MemorySegment.class) {
-            return AddressLayout.class;
+            return CD_AddressLayout;
         } else {
             throw new IllegalStateException("Unknown type: " + type);
         }
     }
 
-    private void emitInvokeStatic(Class<?> owner, String methodName, String descriptor) {
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:940]: It is great to see these ad-hoc routines (most of which exist in one form or another in all our JDK code generators) go away!

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java [hunk 3]
```diff
@@ -65,69 +63,67 @@
...
-            H_INVOKESTATIC,
-            METHOD_HANDLES_INTRN,
-            "classData",
-            CLASS_DATA_DESC,
-            false);
-    private static final ConstantDynamic CLASS_DATA_CONDY = new ConstantDynamic(
-            ConstantDescs.DEFAULT_NAME,
-            OBJECT_DESC,
-            BSM_CLASS_DATA);
+    private static final ClassDesc CD_Arena = desc(Arena.class);
```
Comment (mcimadamore) [inline src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:77]: I love that we're able to describe what we need to use at a higher-level of abstraction than just strings.

**Review by mcimadamore — APPROVED** [review]
Looks great!

File: src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java (hunk 1 above)
Comment (asotona) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Stackmap generator does not perform full verification, it only performs fast pass through the code and hits only errors preventing to construct valid stack maps (as for example stack underflow).
Verifyier on the other hand does full verification similar to when the class is loaded.

Comment (asotona) [reply src/java.base/share/classes/jdk/internal/foreign/abi/BindingSpecializer.java:216]: Classfile API allows to pass-through complete class members or attributes, so resulting class might be invalid even the particular code and stack maps have been generated. 

This is brief list of verifications performed when `verify` is explicitly called:
- class version
- signatures
- instruction opcodes
- CP entries matching instructions (matching types, array dimensions limits, etc...)
- stack map frames and assignability between reference types
- flow control (branches, exception handlers, falling through code end, falling through initialised, etc..)
- exception tables (offsets, throwing Throwables, etc...)
- local variable tables 
- switches (low/high, number of keys, etc...)
- method calls (arguments, operands, constructor call target, constructors returning void, etc...)
- return values

**Comment by JornVernee** [comment]
Thanks for the reviews. I've now re-implemented the `PERFORM_VERIFICATION` flag as well, using `ClassModel::verify`.

**Pull request closed without merging**